File: grades/grades_test.go
package grades

import "testing"

func TestV(t *testing.T) {
	cases := []struct {
		difficulty int
		want       int
		ok         bool
	}{
		{1, 0, true},   // 1a/V0
		{12, 0, true},  // 4c/V0
		{13, 1, true},  // 5a/V1
		{15, 2, true},  // 5c/V2
		{18, 4, true},  // 6b/V4
		{22, 6, true},  // 7a/V6
		{23, 7, true},  // 7a+/V7
		{27, 10, true}, // 7c+/V10
		{39, 22, true}, // 9c+/V22
		{0, 0, false},
		{40, 0, false},
	}
	for _, c := range cases {
		got, ok := V(c.difficulty)
		if got != c.want || ok != c.ok {
			t.Errorf("V(%d) = %d,%v want %d,%v", c.difficulty, got, ok, c.want, c.ok)
		}
	}
}

func TestVFromDisplay(t *testing.T) {
	if v, ok := VFromDisplay(18.4); !ok || v != 4 {
		t.Errorf("VFromDisplay(18.4) = %d,%v want 4,true", v, ok)
	}
	if v, ok := VFromDisplay(22.6); !ok || v != 7 {
		t.Errorf("VFromDisplay(22.6) = %d,%v want 7,true (rounds to 23)", v, ok)
	}
}

File: packages/core/src/grades.test.ts
import { describe, expect, it } from "vitest";
import { fontFromV, v, vFromDisplay, vFromFont } from "./grades";

describe("v", () => {
  const cases: Array<[difficulty: number, want: number | undefined]> = [
    [1, 0],
    [12, 0],
    [13, 1],
    [15, 2],
    [18, 4],
    [22, 6],
    [23, 7],
    [27, 10],
    [39, 22],
    [0, undefined],
    [40, undefined],
  ];

  it.each(cases)("v(%i) = %s", (difficulty, want) => {
    expect(v(difficulty)).toBe(want);
  });
});

describe("vFromDisplay", () => {
  it("truncates within a grade band", () => {
    expect(vFromDisplay(18.4)).toBe(4);
  });

  it("rounds up across a band boundary", () => {
    expect(vFromDisplay(22.6)).toBe(7);
  });
});

describe("vFromFont", () => {
  const cases: Array<[font: string, want: number | undefined]> = [
    ["4", 0],
    ["5", 1],
    ["5+", 2],
    ["6A", 3],
    ["6a+", 3],
    ["6B+", 4],
    ["6C", 5],
    ["7A", 6],
    ["7a+", 7],
    ["7B", 8],
    ["7C+", 10],
    ["8A", 11],
    ["8b+", 14],
    ["9A", 17],
    [" 6b ", 4],
    ["6D", undefined],
    ["V5", undefined],
    ["", undefined],
  ];

  it.each(cases)("vFromFont(%j) = %s", (font, want) => {
    expect(vFromFont(font)).toBe(want);
  });
});

describe("fontFromV", () => {
  it("round-trips through vFromFont for every V grade", () => {
    for (let grade = 0; grade <= 17; grade++) {
      const font = fontFromV(grade);
      expect(font).toBeDefined();
      expect(vFromFont(font!)).toBe(grade);
    }
  });

  it("is undefined outside the boulder range", () => {
    expect(fontFromV(-1)).toBeUndefined();
    expect(fontFromV(18)).toBeUndefined();
  });
});
